### DISCUSSION TITLE
Question about bottlenecks in libvips

### DISCUSSION THREAD

**Discussion by dv1** [libvips/libvips#4310] [Q&A]
I tried to load and render a very large image, and it took `vips_sink_screen` a rather long time to finish. I saw that the `vips_region_prepare` calls were the bottlenecks (each one taking ~30ms).

`vipsdisp-tiny` and `vipsdisp` both do the rendering much faster. But then I saw that this is because both use `vips_subsample`. If I disable subsampling, they both also suffer from the same issue. Interestingly, it did not matter if I placed the `vips_subsample` call before or after calls like `vips_colourspace`.

This makes me wonder what is going on. Since the order does not seem to matter, it does not seem to me that it is a processing bottleneck - processing was fast with subsampling applied, no matter if subsampling in the beginning or in the end. And, in the no-subsampling case, when I copied the pixels from the memory buffer prepared by `vips_region_prepare`, my manual pixel copy (a for-loop copying rows with `memcpy`) was much faster than 30ms. So what is vips doing? I can only imagine that the subsampling helps performance here because it ends up fetching fewer pixels from the source, which ultimately means that whatever image loading library is used only has to fetch a subset of the image pixels.

**Comment by jcupitt** [comment]
Hello again,

It's a tricky problem, and speed depends (a lot!) on the exact sequence of calls you are using and the details of the underlying image format.

You need an image pyramid for really fast zoom and scroll, so you'll need to check the exact format and use the pyramidal path for files like that. vipsdisp has a lot of code for this. Try it with one of the openslide formats, for example:

https://openslide.cs.cmu.edu/download/openslide-testdata/Aperio/CMU-2.svs

That's an 80k x 30k pixel image and it opens and zooms instantly. Because it's using the precomputed pyramid, you'll find there's little aliasing, despite the use of `subsample` for intermediate layers.

If there's no pyramid, then you need to shrink on the fly. A high-quality shrink (eg. `resize`) will need to scan the *entire* image. For example, a 30k x 30k pixel JPEG will be decompressed to a 30,000 x 30,000 pixel temporary file (almost 4gb), then `resize` will have to read the entire 4gb to generate a 500x500 pixel overview. This will take a long time (obv).

If you use `subsample`, zooming out by a factor of 60 (30000/500) means you'll only read every 60th scanline, so it'll run 60x faster.

vipsdisp tries to cache and reuse tiles, so it'll keep the 500x500 layer in memory and use that as the low-res basis for new tiles as you zoom in. It has a sparse tile pyramid, so it'll select the best mix of available images when composing each frame.

Did you try debug mode? In vipsdisp, press ^D and it'll toggle a visualisation of the sparse tile pyramid so you can watch tiles being swapped in and out.

One thing you could do (which vipsdisp does not) is to do an initial render with `subsample` and then follow up with a high-quality `resize` in the background. This can get a bit hairy though, and I don't think the effort is worth it myself for my imaging work. If I care about high quality display, I use a pyramidal file format.

**Reply by dv1** [reply]
Ah, so `subsample` essentially does a nearest neighbour scaling by reading a fraction of the data, as I wrote before. Makes sense.

From what I've seen, `vipsdisp` creates the pyramid by creating a `VipsImage` per level. For each level, it caches tiles by calling `vips_region_prepare` once `vips_sink_screen` reports the tile as ready (through the notification callback) and then stores the tile (or the subsection that is still within the image in case of tiles at the border) in a GTK texture. The actual drawing is then done by rendering said textures. Care must be taken to align the textures such that the borders between them are seamless. Correct?

**Reply by jcupitt** [reply]
That's right. You have to fetch each tile twice: first to queue a request, and a second time to fetch the computed pixels.

The tile alignment is a bit busted right now, I need to fix it.

**Reply by dv1** [reply]
Regarding tile alignment: I suggest to add to the `vips_sink_screen` API documentation that it is guaranteed that:

1. Tiles never overlap
2. Tiles always are of the specified tile width and height
3. Tiles always start at the image origin

Because then, it becomes possible to generate a tile index out of the tile coordinates. For example, with configured tile size being 256x256 pixels, it becomes possible to deduce that a tile that starts at coordinates 512,256 is at tile row 1 and tile column 2. Out of that, one can then create a tile index by calculating: `tile_index = tile_column + tile_row * round_up(image_width/tile_width)`. This index is then useful for assigning unique IDs to tiles per pyramid level - a valuable property for a tile cache.

It then also becomes easy to associate tiles of different levels in the pyramid, and the levels use a classical subdivision by halving both width and height. For example, suppose there are two levels, and level 1 is the full-detail level. This then means that 1 tile from level 0 covers the same area as 4 tiles. Then, tile 1 at level 1 will cover tiles 2, 3 of the first row and tiles 2, 3 of the second row on level 0. This is then useful for composite images where lower level, less detailed tiles can still be shown when zoomed in until the higher detailed higher level tiles fully load. In this case, tile 1 from level 0 can cover the area until the four more detailed tiles fully load.

**Reply by jcupitt** [reply]
Yes, we could add a note about this. Would you like to make a PR?
